### PR TITLE
nicer errors from assert_unsafe_precondition

### PR DESCRIPTION
**Pull request by RalfJung**
This makes the errors shown by cargo-careful nicer, and since `panic_no_unwind` is `nounwind noreturn` it hopefully doesn't have bad codegen impact. Thanks to @bjorn3 for the hint!

Would be nice if we could somehow supply our own (static) message to print, currently it always prints `panic in a function that cannot unwind`. But still, this is better than before.

### CODE REVIEW

**Comment by RalfJung** [comment]
I extended this to a slight refactor of the no-unwind part of the core::panicking module, so that we can get a nicer error message.
r? @bjorn3  since this is now mostly a compiler change.

File: library/core/src/panicking.rs [hunk 1]
```diff
@@ -84,12 +84,28 @@ fn panic_bounds_check(index: usize, len: usize) -> ! {
...
+
+/// Like panic_fmt, but without unwinding and track_caller to reduce the impact on codesize.
+#[cold]
+// If panic_immediate_abort, inline the abort call,
+// otherwise avoid inlining because of it is cold path.
+#[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
+#[cfg_attr(feature = "panic_immediate_abort", inline)]
+#[cfg_attr(not(bootstrap), rustc_nounwind)]
+#[cfg_attr(bootstrap, rustc_allocator_nounwind)]
+pub fn panic_fmt_nounwind(fmt: fmt::Arguments<'_>) -> ! {
```
Comment (RalfJung) [inline library/core/src/panicking.rs:108]: I hope I got all these attributes right. :D They are basically copied from `panic_fmt`.

**Comment by RalfJung** [comment]
Ah, looks like that codegen test wants things to be inlined when `panic_fmt_nounwind` is called from `panic_no_unwind`...

**Comment by RalfJung** [comment]
... but of course when we call it elsewhere we don't want it to be inlined. There's no way to force inlining for one but not all call sites, is there?

File: library/core/src/intrinsics.rs [hunk 2]
```diff
@@ -2210,8 +2210,8 @@ macro_rules! assert_unsafe_precondition {
             #[inline(always)]
             fn runtime$(<$($tt)*>)?($($i:$ty),*) {
                 if !$e {
-                    // abort instead of panicking to reduce impact on code size
-                    ::core::intrinsics::abort();
+                    // don't unwind to reduce impact on code size
+                    ::core::panicking::panic_fmt_nounwind(format_args!("unsafe precondition violated"));
```
Comment (saethlin) [inline library/core/src/intrinsics.rs:2214]: If you give a moose a muffin...

Can we stringify the predicate and pass that instead?

Comment (RalfJung) [reply library/core/src/intrinsics.rs:2214]: I think I'd be in favor of letting the macro user give a string. But that's for a future PR, first this one needs to work.

**Comment by RalfJung** [comment]
Yeah I am out of ideas here for how to make that test pass...

**Comment by RalfJung** [comment]
@bjorn3 any idea how to make progress here?

**Comment by bjorn3** [comment]
Maybe just accept code duplication between the two functions?

**Comment by RalfJung** [comment]
Is that really the beet we can do? Sounds terrible.^^

**Comment by RalfJung** [comment]
Cc @nikic @bbjornse  who worked on this test in the past -- we're having trouble with the `assembly/stack-protector/stack-protector-heuristics-effect.rs` test failing and I have no idea what this is even testing or why it fails now.

**Comment by nikic** [comment]
I think the actual problem here is that this implementation completely fails the goal of reducing code size impact in the caller: You are passing in a `fmt::Arguments`, which is a complex structure that requires a stack allocation. What you should be doing is only pass in a `&str` and then translate that to the formatting machinery internally.

This is probably also how it ends up failing the stack protector test.

**Comment by RalfJung** [comment]
I shouldn't be passing anything if that all gets inlined. And both the code before and after this PR construct a `fmt::Arguments` in the end...

... so, I do not understand how this can make a difference. But I'll try your suggestion.

**Comment by nikic** [comment]
> I shouldn't be passing anything if that all gets inlined. And both the code before and after this PR construct a `fmt::Arguments` in the end...

But we don't *want* this function to be inlined, right? In fact, it is marked as `#[inline(never)]` (outside panic_immediate_abort).

**Comment by RalfJung** [comment]
Oh wait `panic_no_unwind` was meant to call the `_inline` version of the function of course...

... or are you saying the error is probably due to the other change, for the debug assertion, since this actually is a stdlib with debug assertions and we are now passing a rather large argument to `panic_fmt_nounwind`? That is possible of course.

**Comment by nikic** [comment]
> ... or are you saying the error is probably due to the other change, for the debug assertion, since this actually is a stdlib with debug assertions and we are now passing a rather large argument to `panic_fmt_nounwind`? That is possible of course.

That :)

**Comment by RalfJung** [comment]
Aha, that did it. :)
@bjorn3 should be ready for review.

File: library/core/src/panicking.rs [hunk 3]
```diff
@@ -29,6 +29,67 @@
...
+#[cold]
+// If panic_immediate_abort, inline the abort call,
+// otherwise avoid inlining because of it is cold path.
+#[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
+#[cfg_attr(feature = "panic_immediate_abort", inline)]
+#[track_caller]
+#[lang = "panic_fmt"] // needed for const-evaluated panics
+#[rustc_do_not_const_check] // hooked by const-eval
+#[rustc_const_unstable(feature = "core_panic", issue = "none")]
+pub const fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
```
Comment (bjorn3) [inline library/core/src/panicking.rs:47]: Could you move this back to where it was? The moves makes the diff harder to check.

File: library/core/src/panicking.rs [hunk 4]
```diff
@@ -84,62 +145,16 @@ fn panic_bounds_check(index: usize, len: usize) -> ! {
...
+/// Panic because we cannot unwind out of a function.
+///
+/// This function is called directly by the codegen backend, and must not have
+/// any extra arguments (including those synthesized by track_caller).
 #[cold]
-#[inline(never)]
 #[lang = "panic_no_unwind"] // needed by codegen for panic in nounwind function
+#[cfg_attr(not(bootstrap), rustc_nounwind)]
+#[cfg_attr(bootstrap, rustc_allocator_nounwind)]
 fn panic_no_unwind() -> ! {
```
Comment (bjorn3) [inline library/core/src/panicking.rs:156]: And maybe put this before `panic_str_nounwind`?

**Comment by bjorn3** [comment]
I think it looks fine, but the above suggestions will make it easier for me to check that this is indeed the case.

File: library/core/src/panicking.rs (hunk 4 above)
Comment (RalfJung) [reply library/core/src/panicking.rs:156]: This is a small helper wrapper like e.g. the bounds-check wrapper, we have a bunch of those so I figured they should go together.

File: library/core/src/panicking.rs (hunk 3 above)
Comment (bjorn3) [reply library/core/src/panicking.rs:47]: The current order is not entirely random. It puts every caller before every callee, so you can read from top to bottom to get the order in which functions are called.

Comment (RalfJung) [reply library/core/src/panicking.rs:47]: Oh, that explains why it is so hard to read... I find it much easier when the functions are defined before they are used. I have to scroll up and down all the time right now, since code keeps calling functions that are only defined (and documented) much later.

Also, `const_panic_fmt` is after `panic_str` despite calling that, so the order is not entirely the way you say it is.

Comment (RalfJung) [reply library/core/src/panicking.rs:50]: All right, the reordering is a separate commit now.

I still feel like defining the two functions doing the main work first makes it a lot easier to read the rest of the file -- they establish the "base vocabulary" that everything else is using. Right now I find myself having to speculate how it all holds together until I finally reach `panic_fmt`, *then* I see what really happens, and then I have to go back to the top and re-read everything.

Basically: terms should be defined before they are used. This is obviously true when doing technical writing (I cannot first use a technical term and then 2 pages later define it for the first time). And IMO it is just as true in source code, where "terms" includes things like functions and types that the code uses to express what it does.

**Comment by matthiaskrgr** [comment]
@bors r-
I guess this makes these tests fail, since the pr changed some panicking-related code? :thinking: 
https://github.com/rust-lang/rust/pull/102816#issuecomment-1272401936

**Comment by RalfJung** [comment]
Yeah it is probably this PR but I don't understand why...
```
   = note: "cc" "-m64" "/tmp/rustcSwOBub/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/extern/issue-64655-extern-rust-must-allow-unwind.fat0/a.issue_64655_extern_rust_must_allow_unwind.2a84b153-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/extern/issue-64655-extern-rust-must-allow-unwind.fat0/auxiliary" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-c9b03ebf09765622.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/extern/issue-64655-extern-rust-must-allow-unwind.fat0/a" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--enable-new-dtags" "-Wl,-z,origin"
   = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-c9b03ebf09765622.rlib(compiler_builtins-c9b03ebf09765622.compiler_builtins.580e2b8a-cgu.111.rcgu.o): in function `core::hint::unreachable_unchecked':
           compiler_builtins.580e2b8a-cgu.111:(.text._ZN4core4hint21unreachable_unchecked17h9bac5ed06aa3f3efE+0xf): undefined reference to `core::panicking::panic_str_nounwind'
           collect2: error: ld returned 1 exit status
```
Why would `panic_str_nounwind` suddenly not exist? I think I once saw a similar error locally when I marked `panic_no_unwind` as `inline`. Is there some pre-existing bug where the collector does not correctly mark `panic_no_unwind` as needing codegen, or so?

**Comment by RalfJung** [comment]
Maybe making `panic_no_unwind` pub helps -- after all, arbitrary crates *do* have direct calls to this function, the calls just happen in a way that circumvents visibility checking.

@bors r=bjorn3 rollup=never

**Comment by RalfJung** [comment]
Oh funny, now it also fails in the PR. Is this a recently added test?
@bors r-

**Comment by RalfJung** [comment]
`panic_str_nounwind` is already a `pub` function so the only way I see to fix this is to remove its `inline` attribute.

File: library/core/src/panicking.rs [hunk 5]
```diff
@@ -84,62 +149,16 @@ fn panic_bounds_check(index: usize, len: usize) -> ! {
...
 }
 
-// This function is called directly by the codegen backend, and must not have
-// any extra arguments (including those synthesized by track_caller).
+/// Panic because we cannot unwind out of a function.
+///
+/// This function is called directly by the codegen backend, and must not have
+/// any extra arguments (including those synthesized by track_caller).
 #[cold]
-#[inline(never)]
```
Comment (bjorn3) [inline library/core/src/panicking.rs:90]: Maybe re-adding this will work? Also please add the inline attributes of panic_str_nounwind back if that works.

**Comment by RalfJung** [comment]
Oh wait, this is not about `panic_no_unwind`...
```
   = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-c9b03ebf09765622.rlib(compiler_builtins-c9b03ebf09765622.compiler_builtins.9e069ada-cgu.52.rcgu.o): in function `core::hint::unreachable_unchecked':
           compiler_builtins.9e069ada-cgu.52:(.text._ZN4core4hint21unreachable_unchecked17ha5db74bd963fb912E+0xf): undefined reference to `core::panicking::panic_str_nounwind'
           collect2: error: ld returned 1 exit status
```
It's about the "usual" problem that compiler-builtins is very special, and somehow it calling `panic_str_nounwind` via unreachable_unchecked does not work.

**Comment by RalfJung** [comment]
Ah that also explains why this was not caught in CI earlier -- unreachable_unchecked using assert_unsafe_precondition is a recent change.

**Comment by RalfJung** [comment]
@bors r-
let's land https://github.com/rust-lang/rust/pull/102894 first

**Comment by nnethercote** [comment]
`diesel` has been noisy lately.

@rustbot label: +perf-regression-triaged
